proposals/p2240.md

```diff
@@ -0,0 +1,758 @@
+ways it would be very desirable to make their Carbon counterparts postfix as
+well. Notably, it would let us describe variadic evaluation very cleanly in
+terms of replacing `[:]` with `[$I]`. However, this turns out to be an awkward
+fit for Carbon. Consider this example:
+
+```
+fn SumInts(..., [:]params: [i64;]) -> i64
+```
+
+If `[:]` and `...,` were suffix operators, this would instead be:
+
+```
+fn SumInts(params: [i64;][:] ...,) -> i64
+```
+
+But this is ambiguous about whether the `[:]` operator applies to
+`params: [i64;]` or just `[i64;]`, so we need to add a set of parentheses:
+
+```
+fn SumInts((params: [i64;])[:] ...,) -> i64
+```
+
+To state the point more generally: when an expression mixes prefix and suffix
+operators, their order is determined by precedence and parentheses, which is
+usually less readable than relying on the linear order of operators with the
+same fixity. The binding operator `:` is effectively a prefix operator (because
+its left-hand side is a single token), so if `[:]` is a suffix operator, code
+that mixes the two will be somewhat harder to read. That's a particular problem
+because, as our motivating examples illustrate, variadic function signatures
+will often have those two operators in close conjunction, and readability is at
+a premium in function signatures.
+
+C solved a very similar problem with mixing prefix `*` and postfix `.` by
+introducing a separate `->` operator. We could in principle solve this issue the
+same way, introducing a separate "variadic binding operator" `$`, such that
+`params$ [i64;]` is equivalent to `(params: [i64;])[:]`. However, this would
+trade a readability problem for a teachability problem, by introducing more
+syntax that programmers need to learn before they can even read variadic
+function signatures.
+
+### Keyword syntax
+
+> **TODO:** Decide whether to make this the primary proposal, and restructure
+> the discussion as appropriate.
+
+We could build the syntax around keywords rather than symbolic tokens, much like
+the
+[Swift proposal](https://github.com/apple/swift-evolution/blob/main/proposals/0393-parameter-packs.md).
+For concreteness, the keywords I will discuss here are as follows:
+
+| Symbolic token | Keyword replacement               |
+| -------------- | --------------------------------- |
+| `[:]`          | `expand`                          |
+| `...,`         | `repeat` and `unpack` (see below) |
+| `...and`       | `all_of`                          |
+| `...or`        | `any_of`                          |
+| `...{`         | `repeat {`                        |
+
+This approach seems to work best in combination with some other nominally
+independent changes:
+
+-   Forbid `expand` in pattern contexts.
+-   Allow `repeat` only in expression contexts and `unpack` only in pattern
+    contexts.
+-   Require any usage of a pack-type variable to be prefixed with a new keyword
+    `each`.
```

Comment:
@zygoloid says: this gives us syntactic room to give the pack itself an API, that can be accessed by referring to the pack without `each`. For example, a templated pack could support indexing (or maybe: indexing into a pack is a template operation):

```
fn ReverseTuple[template pack T:! type](x: (repeat each T))
    -> (repeat[i: i32] T[sizeof(T) - i - 1]) {
  return (repeat[i: i32] x[sizeof(T) - i - 1]);
}
```

Reply:
(Expectation here is that `template` is needed to do pack indexing, since we do not expect to be able to type check these generically.)

We also discussed potentially getting the elements of the pack
```
fn F[template pack T:! type](x: (repeat each T)) {
    repeat[U in T, i: i32] {
        match (x) {
            // matches if every element of the tuple has the same type
            case (pack a: U) => { }
            // would match always
            case (pack a: each T) => { }
            // same as previous case
            case (repeat[V in T] a: V) => { }
        }
    }
}
```